### ISSUE TITLE
Negative habits misrepresented in the Weekly report

### ISSUE THREAD

**Issue by lynxlynxlynx** [HabitRPG/habitrpg-email-server#5]
Habits that are negative-only can't get blue, so their maximum is at the yellow value range. However, the counter in the weekly report still considers such habits as bad, even though this is the best they can be (besides not existing).

I propose that they (and only they) be counted as good if their value is above -1.1. The asymptote seems to be near -1.0xyz for my few examples, not zero as I expected. The code is currently just checking if the value is less than 1 and disregarding the type.

**Comment by paglias** [comment]
@lynxlynxlynx thanks for the issue and the PR. I'll check if the -1.1 value is correct and will merge your PR

**Comment by lynxlynxlynx** [comment]
The exact value is about -1.02666 (many digits, looks like some floating point bruhaha) and matches across tasks (over a span of 7mo - as far as my graph goes), even when pushed lower in between. The closest approximate fractions are nothing particularly nice.

**Comment by paglias** [comment]
Yes, I'm trying to find out if we have a more exact way of getting that value than just using a fixed number, probably it's based on some specific variables

**Comment by Alys** [comment]
When you create a new Habit, its value is 0. If you click the negative button once, its value becomes -1.026657310999762 (and its colour becomes orange). Click it twice and it becomes -2.081442996376489. So with `weakHabitCutoff = -1.1`, we're saying that a negative-only Habit is weak if you've clicked it twice or more, but it's strong if you've clicked it only once or not at all. Is that the intention? Would it make more sense to be strong only if you've never clicked it at all (value > -1.02) ?

When cron runs, the value of a negative-only Habit is halved, so if you clicked a Habit once yesterday (value becomes -1.026657310999762, orange), then at the start of today it becomes -0.513328655499881 and its colour becomes yellow. I would suggest that the Habit could at that point be considered strong.

If `weakHabitCutoff` is changed to `-1.02` then the behaviour would match what I think is intuitive, but that's just my opinion of "intuitive". :) Is there a case for saying that a negative-only Habit is still strong if you've clicked it once?

**Comment by lynxlynxlynx** [comment]
It may be a separate bug, but it is like I noted above, my values aren't changing (halving) anymore, not asymptotically approaching 0. The ugly number is their glass ceiling. They're also as yellow as a new todo, but that's stating the same thing.

**Comment by Alys** [comment]
When you say the "ugly number", do you mean -1.026657310999762 ?

Regardless of numbers, is the intention that a negative Habit created today is strong only if it's never been clicked?

If you'd like us to investigate the bug you're experiencing, please tell us your User ID. I would like to look into it, if you don't mind me looking at your Habits!

**Comment by lynxlynxlynx** [comment]
The intention was to mark maxed habits as strong, with a buffer to avoid problems and not obscure the code too much. -1 would have been nicely symmetric with the previous check, but is currently not good. Is the algorithm public somewhere, because I doubt I'm the only one seeing this?

**Comment by Alys** [comment]
The algorithm for changing the value of Habits during cron is at https://github.com/HabitRPG/habitrpg/blob/develop/common/script/index.js#L2542
As you can see, it's pretty simple, so I am concerned that it's not working for your account and I would very much like to investigate if you don't mind letting me look at your Habits. If you'd prefer to keep them private, and if you have time, would you be able to export the data for your Habits and then edit the export to remove any private information from the text fields? It's just the values and history that I need to see, so as long as I can identify the Habits that aren't halving properly, I don't need to see their text.

**Comment by Alys** [comment]
@lemoness We should probably get a comment from you about how you'd like this to work. The issue is that negative-only Habits are always reported in the emails as weak, even if they've changed back to yellow and you haven't done the bad Habit for a while. The PR associated with this issue is proposing that a negative-only Habit should be considered strong under a certain condition. That condition could be ONE of these options:

1. The Habit is yellow. This means that you have never clicked it, or that you have not clicked it recently and the Habit has aged back to yellow. This is the strict condition - you must not have done the bad thing at all (recently) for the Habit to be considered strong. (_This is my preference, but I don't feel strongly about it._)

2. The Habit is slightly orange but not very orange. For a new Habit, this means you have clicked it once only. For an older Habit, it means you've clicked it potentially multiple times, but not very often recently so it has aged back to being slightly orange. This is a less strict condition - you are allowed about one click a day at most on the negative Habit and it's still considered strong; it gives you a bit of a buffer before we tell you you are doing poorly at that Habit. One possible problem with this is that some people might have negative-only Habits that can be clicked only once a day at most (e.g., click a junk food Habit if you ate junk any time today); for those people an orange Habit should be considered weak so this condition would not be suitable.

3. Or something else. :)

What would your preference be?

**Comment by lynxlynxlynx** [comment]
Yeah, that looks like a horrible place to carry a bug, but it probably only exposes it. My first guess was the strict boolean check+data corruption.
UID: 10c530d0-a796-476b-8a31-6f048c6e8a8c

**Comment by Alys** [comment]
@lynxlynxlynx Thanks for your ID! I'll have a look at your Habit data, and watch how the values change (or not) over the next couple of days.

@lemoness, I forgot to mention above, we should do the same sort of thing for positive-only Habits, and also probably prevent any weak/strong assignment for Habits that have neither plus nor minus selected, because people use them as separators, notes, inspirational messages, etc. However we could focus on those cases after we've sorted out the negative Habits.

**Comment by lemoness** [comment]
My preference would be for all-yellow negative habits to count as strong, and for all yellow positive-only habits to be still be counted as weak. I also agree that ones with no markers should be excluded.

**Comment by lynxlynxlynx** [comment]
Positive-only is fine already, I think — the threshold is at 1 and first click should get you to that 1.02... I've added the mentioned ignorance to the PR.

**Comment by Alys** [comment]
Oh yes, I expect you're right about positive-only Habits!

The task value cut-offs for the different colours are defined here:
https://github.com/HabitRPG/habitrpg/blob/dcd7cae31295566b8fd08e5a8cec076f975b45b0/common/script/index.js#L396-L410
https://github.com/HabitRPG/habitrpg/blob/dcd7cae31295566b8fd08e5a8cec076f975b45b0/website/public/css/global-colors.styl#L1-L9

**Comment by Alys** [comment]
@lynxlynxlynx I've compared your negative-only Habits' values from yesterday and today and they do seem to be changing as expected. Are there any particular ones that you've noticed aren't ageing correctly?

**Comment by lynxlynxlynx** [comment]
I'm looking at them through the progress graph icon.  Granted, they don't show recent history (december), but for the past that ugly is pervasive. Since the color is right, the bug must be elsewhere indeed — preening.

**Comment by Alys** [comment]
@lynxlynxlynx Can you log that as a separate issue? Please include screenshots that show where the problem is, and explain what the ugly part is.

**Comment by lynxlynxlynx** [comment]
Sure. In the meanwhile, the pr is ready for merging as far as I can tell.

**Comment by paglias** [comment]
@Alys if you think it's ok I'll merge the PR and deploy

**Comment by Alys** [comment]
It's good!
